### ISSUE TITLE
Handle SLURM errors

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-server#343]
I ran a fractal workflow doing cellpose on a big 3D dataset and it ran out of memory on the cpu-mid executor. The following was in the slurm.err file:
```
slurm error contain:
srun: error: vcl1016: task 0: Out Of Memory
slurmstepd: error: Detected 1 oom-kill event(s) in StepId=130.0. Some of your processes may have been killed by the cgroup out-of-memory handler.
```

This also showed up on the server (in the console running the server), so I guess this is a state we're not handling well yet:
```
Exception in thread Thread-434:
Traceback (most recent call last):
  File "/ENV/FOLDER/fractal-server-v1-b4/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/ENV/FOLDER/fractal-server-v1-b4/lib/python3.8/site-packages/cfut/__init__.py", line 61, in run
    self.callback(self.waiting[filename])
  File "/ENV/FOLDER/fractal-server-v1-b4/lib/python3.8/site-packages/fractal_server/app/runner/_slurm/executor.py", line 338, in _completion
    fut.set_exception(exc)
  File "/ENV/FOLDER/fractal-server-v1-b4/lib/python3.8/concurrent/futures/_base.py", line 547, in set_exception
    raise InvalidStateError('{}: {!r}'.format(self._state, self))
concurrent.futures._base.InvalidStateError: CANCELLED: <Future at 0x7f778ce452e0 state=cancelled>
```

**Comment by tcompa** [comment]
What fractal-server currently supports are tasks error, which are then encapsulated in `TaskExecutionError`s. This error is specific of the SLURM backend, but we can have a look and try to be more flexible.

---

Side-comment: good to know that FMI enforces memory hard limits through cgroup. This is not always the case (e.g. it's not like this at UZH), so that the error may be different depending on the SLURM config. My guess is that this would fail "correctly" at UZH, with an out-of-memory error from the task that is captured as a `TaskExecutionError`.

**Comment by tcompa** [comment]
Reminder to self:
This can be tested by having a task that sleeps 10 seconds and setting a walltime limit of 1 second - so that the failure should be a SLURM one.

**Comment by tcompa** [comment]
Another (non-equivalent) option: submit a job to an invalid partition.

**Comment by tcompa** [comment]
For the record,

Errors that take place in the `callback` function within `FileWaitThread` (see [here](https://github.com/sampsyo/clusterfutures/blob/master/cfut/__init__.py)) are not handled, at the moment, because it's a different thread.

Here is a minimal non-working example
```python
import threading
import time
import logging


class worker_thread(threading.Thread):
    def run(self):
        for i in range(2):
            logging.warning(f"Logging from worker_thread, {i}-th iteration")
            time.sleep(1)
        logging.warning("worker_thread will now raise")
        raise RuntimeError("this is an error")


class main_thread():
    def __init__(self):
        t = worker_thread()
        t.start()
        time.sleep(0.5)   # this is just to get nicer output, with separated main/worker logs
        for i in range(5):
            print()
            logging.warning(f"Logging from main_thread, {i}-th iteration")
            logging.warning(f"  Is worker_thread alive? {bool(t.is_alive())}")
            time.sleep(1)

main_thread()
```
with output
```
WARNING:root:Logging from worker_thread, 0-th iteration

WARNING:root:Logging from main_thread, 0-th iteration
WARNING:root:  Is worker_thread alive? True
WARNING:root:Logging from worker_thread, 1-th iteration

WARNING:root:Logging from main_thread, 1-th iteration
WARNING:root:  Is worker_thread alive? True
WARNING:root:worker_thread will now raise
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/home/tommaso/Fractal/test_threads/mnwe.py", line 12, in run
    raise RuntimeError("this is an error")
RuntimeError: this is an error

WARNING:root:Logging from main_thread, 2-th iteration
WARNING:root:  Is worker_thread alive? False

WARNING:root:Logging from main_thread, 3-th iteration
WARNING:root:  Is worker_thread alive? False

WARNING:root:Logging from main_thread, 4-th iteration
WARNING:root:  Is worker_thread alive? False
```

Possibly useful references:
* https://docs.python.org/dev/library/threading.html#threading.excepthook: We can use this to handle uncaught exceptions in threads, but we are not (yet?) able to use it to re-raise.
* https://docs.python.org/dev/library/threading.html#event-objects: We can add an `Event` attribute to the `FractalSlurmExecutor` (in the same way as there is a `FileWaitThread` attribute), and `set()` it when exceptions take place in the `FileWaitThread`, but how would we check for these exceptions in the main thread?
* ...

**Comment by tcompa** [comment]
Apart from #366, which is generic for any error raised within the `FileWaitThread`, we need some checks that are more related to SLURM jobs.
The basic tool in that case should be a function similar to this one
```python
def check_job_status(jobid: int) -> str:
    cmd = f"squeue --noheader --format='%t' --job {jobid}"
    res = subprocess.run(shlex.split(cmd), capture_output=True)
    stdout = res.stdout.decode("utf-8")
    stderr = res.stdout.decode("utf-8")
    if res.returncode != 0 or stderr:
        raise RuntimeError(
                f"check_job_status with {cmd=} failed\n"
                f"{stdout=}, {stderr=}"
                )
    job_status = stdout
    if job_status in ["PD", "R", "CF", "CG", "S"]:
        # SLURM job is running or waiting
        return "wait"
    elif status == "CD":
        # SLURM job completed
        return "go"
    elif status in ["BF", "CA", "F", "TO", "NF", "RV", "SE"]:
        # SLURM job failed
        return "fail"
```

When the `FractalSlurmExecutor.FileWaitThread` checks for the existence of the output pickle file, it should also check the job status (note that the job ID is already known, in that part of the code). Then, it should proceed as follows:
* If the job has not yet completed, nor failed, wait.
* If the job has completed, check for the output pickle file (and fail if it's not there).
* If the job has failed, re-raise something meaningful.

**Comment by tcompa** [comment]
Very relevant discussion and PR:
* https://github.com/sampsyo/clusterfutures/issues/18
* https://github.com/sampsyo/clusterfutures/pull/19

**Comment by tcompa** [comment]
A proof-of-principle is available in #367. If there is an internal failure (i.e. the clusterfutures output file for a given task is not written, e.g. because the SLURM job failed before that), this results in a similar outcome as for a task error, and the `fractal job-status --job-id XXX` command returns something like
```
    job_status_data: {
        'project_id': 1,
        'input_dataset_id': 1,
        'output_dataset_id': 2,
        'workflow_id': 1,
        'overwrite_input': False,
        'worker_init': None,
        'working_dir': '/tmp/pytest-of-tommaso/pytest-58/server_folder0/artifacts/workflow_000001_job_000001',
        'id': 1,
        'start_timestamp': '2022-12-20T11:40:15.126473',
        'status': 'failed',
        'log': (
            'GENERIC ERROR:TRACEBACK:\n'
            "Cluster job 2 finished without writing a result; Original error: [Errno 2] No such file or directory: '/t"
            'mp/pytest-of-tommaso/pytest-58/server_folder0/artifacts/workflow_000001_job_000001/cfut.out.Zb2fjJXEqmrRU'
            "gkattkpE3VxwfYRNRXw.pickle'"
        ),
        'history': None,
    }
```
And we could then make the `log`  field more relevant, e.g. fetching the SLURM-job stderr (similar to what we do with the task errors).

Note that #367 should be considered a proof of principle to see how this could work. For the moment it cannot be merged, because it is based on [an open clusterfutures PR](https://github.com/sampsyo/clusterfutures/pull/19) and [on a custom fork of ours](https://github.com/tcompa/clusterfutures/tree/poll-slurm-mod).

**Comment by tcompa** [comment]
Closed by #379
